node/actors/network/src/gossip/runner.rs

```diff
@@ -181,42 +197,51 @@ impl Network {
                 .add_server(ctx, &*self.block_store, self.cfg.rpc.get_block_rate)
                 .add_client(&get_batch_client)
                 .add_server(ctx, &*self.batch_store, self.cfg.rpc.get_batch_rate)
-                .add_server(ctx, rpc::ping::Server, rpc::ping::RATE);
-
-            // If there is an attester committee then
-            if self.genesis().attesters.as_ref().is_some() {
-                let push_batch_votes_client = rpc::Client::<rpc::push_batch_votes::Rpc>::new(
+                .add_server(ctx, rpc::ping::Server, rpc::ping::RATE)
+                .add_client(&push_batch_votes_client)
+                .add_server::<rpc::push_batch_votes::Rpc>(
                     ctx,
+                    &push_server,
                     self.cfg.rpc.push_batch_votes_rate,
                 );
-                service = service
-                    .add_client(&push_batch_votes_client)
-                    .add_server::<rpc::push_batch_votes::Rpc>(
-                        ctx,
-                        &push_server,
-                        self.cfg.rpc.push_batch_votes_rate,
-                    );
-                // Push L1 batch votes updates to peer.
-                s.spawn::<()>(async {
-                    let push_batch_votes_client = push_batch_votes_client;
-                    // Snapshot of the batches when we last pushed to the peer.
-                    let mut old = BatchVotes::default();
-                    // Subscribe to what we know about the state of the whole network.
-                    let mut sub = self.batch_votes.subscribe();
-                    sub.mark_changed();
-                    loop {
-                        let new = sync::changed(ctx, &mut sub).await?.clone();
-                        // Get the *new* votes, which haven't been pushed before.
-                        let diff = new.get_newer(&old);
-                        if diff.is_empty() {
-                            continue;
+
+            // Push L1 batch votes updates to peer.
+            s.spawn::<()>(async {
+                let push_batch_votes_client = push_batch_votes_client;
+                // Subscribe to what we know about the state of the whole network.
+                let mut recv = self.attestation_state.subscribe();
+                loop {
+                    let diff = recv.wait_for_diff(ctx).await?;
+                    let req = rpc::push_batch_votes::Req {
+                        // If the config has changed, we need to re-request all the votes
+                        // from peer that we might have ignored earlier.
+                        want_snapshot: Some(diff.config_changed),
```

Comment:
I thought about it, considered that a premature optimization. But if you want it, I'll just add it.

Reply:
In practice it will probably not matter that much, if we only have a handful of attesters and a handful of gossip peers. I only thought it everyone added the main node as a static gossip peer, and it's the one announcing the new status (the first to have it), then it will probably always get all the old votes. 

OTOH I like how `want_snapshot` is just a boolean flag. Let's leave it like this and in the future we can add another field to carry the number, and decide how to interpret it.

Reply:
I already updated it. Now it just sends the batch it wants the votes for.